### ISSUE TITLE
Unexpected behaviours on `--keyring-backend test` parameters

### ISSUE THREAD

**Issue by yijiasu-crypto** [tharsis/ethermint#238]
__System info:__ 
```
commit 3a398d
OS: macOS 11.2.3 (20D91)
go version go1.16.5 darwin/amd64
```

__Steps to reproduce:__

1. Start the `Ethermintd` with `--keyring-backend test` parameters

```

KEY="mykey"
CHAINID="ethermint-2"
MONIKER="localtestnet"
MNEMONIC="yard similar hotel exercise calm cousin forget wisdom swallow fatal afraid what dog panther nose age ramp portion floor scene cruise soul strong rose"
# remove existing daemon and client
rm -rf ~/.ethermintd*

make install

ethermintd config keyring-backend test
ethermintd config chain-id $CHAINID

# if $KEY exists it should be deleted
echo $MNEMONIC | ethermintd keys add $KEY --keyring-backend test --algo "eth_secp256k1" --recover

# Set moniker and chain-id for Ethermint (Moniker can be anything, chain-id must be an integer)
ethermintd init $MONIKER --chain-id $CHAINID 

# Change parameter token denominations to aphoton
# cat $HOME/.ethermintd/config/genesis.json | jq '.app_state["staking"]["params"]["bond_denom"]="aphoton"' > $HOME/.ethermintd/config/tmp_genesis.json && mv $HOME/.ethermintd/config/tmp_genesis.json $HOME/.ethermintd/config/genesis.json
# cat $HOME/.ethermintd/config/genesis.json | jq '.app_state["crisis"]["constant_fee"]["denom"]="aphoton"' > $HOME/.ethermintd/config/tmp_genesis.json && mv $HOME/.ethermintd/config/tmp_genesis.json $HOME/.ethermintd/config/genesis.json
# cat $HOME/.ethermintd/config/genesis.json | jq '.app_state["gov"]["deposit_params"]["min_deposit"][0]["denom"]="aphoton"' > $HOME/.ethermintd/config/tmp_genesis.json && mv $HOME/.ethermintd/config/tmp_genesis.json $HOME/.ethermintd/config/genesis.json
# cat $HOME/.ethermintd/config/genesis.json | jq '.app_state["mint"]["params"]["mint_denom"]="aphoton"' > $HOME/.ethermintd/config/tmp_genesis.json && mv $HOME/.ethermintd/config/tmp_genesis.json $HOME/.ethermintd/config/genesis.json

# increase block time (?)
cat $HOME/.ethermintd/config/genesis.json | jq '.consensus_params["block"]["time_iota_ms"]="30000"' > $HOME/.ethermintd/config/tmp_genesis.json && mv $HOME/.ethermintd/config/tmp_genesis.json $HOME/.ethermintd/config/genesis.json

# disable produce empty block
if [[ "$OSTYPE" == "darwin"* ]]; then
    sed -i '' 's/create_empty_blocks = true/create_empty_blocks = false/g' $HOME/.ethermintd/config/config.toml
  else
    sed -i 's/create_empty_blocks = true/create_empty_blocks = false/g' $HOME/.ethermintd/config/config.toml
fi

if [[ $1 == "pending" ]]; then
  if [[ "$OSTYPE" == "darwin"* ]]; then
      sed -i '' 's/create_empty_blocks_interval = "0s"/create_empty_blocks_interval = "30s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_propose = "3s"/timeout_propose = "30s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_propose_delta = "500ms"/timeout_propose_delta = "5s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_prevote = "1s"/timeout_prevote = "10s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_prevote_delta = "500ms"/timeout_prevote_delta = "5s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_precommit = "1s"/timeout_precommit = "10s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_precommit_delta = "500ms"/timeout_precommit_delta = "5s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_commit = "5s"/timeout_commit = "150s"/g' $HOME/.ethermintd/config/config.toml
      sed -i '' 's/timeout_broadcast_tx_commit = "10s"/timeout_broadcast_tx_commit = "150s"/g' $HOME/.ethermintd/config/config.toml
  else
      sed -i 's/create_empty_blocks_interval = "0s"/create_empty_blocks_interval = "30s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_propose = "3s"/timeout_propose = "30s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_propose_delta = "500ms"/timeout_propose_delta = "5s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_prevote = "1s"/timeout_prevote = "10s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_prevote_delta = "500ms"/timeout_prevote_delta = "5s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_precommit = "1s"/timeout_precommit = "10s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_precommit_delta = "500ms"/timeout_precommit_delta = "5s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_commit = "5s"/timeout_commit = "150s"/g' $HOME/.ethermintd/config/config.toml
      sed -i 's/timeout_broadcast_tx_commit = "10s"/timeout_broadcast_tx_commit = "150s"/g' $HOME/.ethermintd/config/config.toml
  fi
fi

# Allocate genesis accounts (cosmos formatted addresses)
ethermintd add-genesis-account "$(ethermintd keys show $KEY -a --keyring-backend test)" 1000000000000000000000aphoton,1000000000000000000stake --keyring-backend test

# Sign genesis transaction
#ethermintd gentx $KEY 1000000000000000000000aphoton --keyring-backend test --chain-id $CHAINID

ethermintd gentx $KEY 1000000000000000000stake --amount=1000000000000000000000aphoton --chain-id $CHAINID --keyring-backend test

# Collect genesis tx
ethermintd collect-gentxs

# Run this to ensure everything worked and that the genesis file is setup correctly
ethermintd validate-genesis

if [[ $1 == "pending" ]]; then
  echo "pending mode is on, please wait for the first block committed."
fi

# Start the node (remove the --pruning=nothing flag if historical queries are not needed)
#ethermintd start --pruning=nothing --rpc.unsafe --trace --log_level "info"
ethermintd start --pruning=nothing --rpc.unsafe --keyring-backend test --trace --log_level "info"
```

2. Run the test cases in this Repo: 
* clone: `https://github.com/monacohq/crypto.com-evm-chain-test`
* `yarn test:ethermint test/0_test_contracts.test.js`

3. Remove the  `--keyring-backend test` parameter, start the ethermintd node again
* (The script will do `rm -rf ~/.ethermintd*` to clear the blockchain state, so we can ensure it's a complete new instance)

4. Re-run the above test

__Expected behavior:__ 

Two rounds of tests should be all passed **OR**
First round of test ALL PASSED and second round of test ALL FAILED

__Actual behavior:__

**At first round of tests, it passed.**

```
➜  eth-call-test git:(master) ✗ yarn test:ethermint
yarn run v1.22.10
$ yarn truffle test --network ethermint
$ /Users/yijiasu/workspace/eth-call-test/node_modules/.bin/truffle test --network ethermint
Warning: ENV_VAR MNEMONIC is not set! Using the default mnemonic!!
Using this MNEMONIC: yard similar hotel exercise calm cousin forget wisdom swallow fatal afraid what dog panther nose age ramp portion floor scene cruise soul strong rose
Using this RPC: 127.0.0.1:8545
Using network 'ethermint'.


Compiling your contracts...
===========================
> Everything is up to date, there is nothing to compile.



  Contract: Test Storage Contract
Using Accounts (10): 
0x90FdB51c13Ce085cE7F9c0dA8683B7327711b064
0xeaCBa4a0B94a7faE0eC4C3E5e303320cB6CAc0AA
0xBd2168238b5cE311B28ABB00BfC119f31E040b86
0x456c710472DAB617fd45264F0971D6be2415ACAB
0x7eFB761519b7fE74be72D303eDe99d81bE7d2813
0x8614b169b22A351BdDb45c150366DEE7Af3B26d6
0xC43383ddB4Cca76d5C0de69c19bCC64cC827fCEC
0x306aa6B89196DBdf917E0d01BC11e86F64A5567d
0xaaca7D818C1623e6C734595017273d50Da8e96fc
0xeE9f52Ba99cfD7f65Ac0a66164eb227db52118A7
==========================

Deployed Storage at: 0xedF82B7bF42006dDD309cB2DEC789316407b3c55
    ✓ should deploy Stroage contract (5299ms)
Stored value 888 by tx: 0x99012e3852607a061ae9815b7304d08901c53be94982dd40764da6b6ccc5dcad
    ✓ should succesfully stored a value (5050ms)
    ✓ should succesfully retrieve a value
```

**At first round of tests, it passed two tests and failed the 3rd test**

```
➜  eth-call-test git:(master) ✗ yarn test:ethermint test/0_test_contracts.test.js 
yarn run v1.22.10
$ yarn truffle test --network ethermint test/0_test_contracts.test.js
$ /Users/yijiasu/workspace/eth-call-test/node_modules/.bin/truffle test --network ethermint test/0_test_contracts.test.js
Warning: ENV_VAR MNEMONIC is not set! Using the default mnemonic!!
Using this MNEMONIC: yard similar hotel exercise calm cousin forget wisdom swallow fatal afraid what dog panther nose age ramp portion floor scene cruise soul strong rose
Using this RPC: 127.0.0.1:8545
Using network 'ethermint'.


Compiling your contracts...
===========================
> Everything is up to date, there is nothing to compile.



  Contract: Test Storage Contract
Using Accounts (10): 
0x90FdB51c13Ce085cE7F9c0dA8683B7327711b064
0xeaCBa4a0B94a7faE0eC4C3E5e303320cB6CAc0AA
0xBd2168238b5cE311B28ABB00BfC119f31E040b86
0x456c710472DAB617fd45264F0971D6be2415ACAB
0x7eFB761519b7fE74be72D303eDe99d81bE7d2813
0x8614b169b22A351BdDb45c150366DEE7Af3B26d6
0xC43383ddB4Cca76d5C0de69c19bCC64cC827fCEC
0x306aa6B89196DBdf917E0d01BC11e86F64A5567d
0xaaca7D818C1623e6C734595017273d50Da8e96fc
0xeE9f52Ba99cfD7f65Ac0a66164eb227db52118A7
==========================

Deployed Storage at: 0xedF82B7bF42006dDD309cB2DEC789316407b3c55
    ✓ should deploy Stroage contract (2271ms)
Stored value 888 by tx: 0x99012e3852607a061ae9815b7304d08901c53be94982dd40764da6b6ccc5dcad
    ✓ should succesfully stored a value (5056ms)
    1) should succesfully retrieve a value
    2) "after each" hook: after test for "should succesfully retrieve a value"


  2 passing (7s)
  2 failing

  1) Contract: Test Storage Contract
       should succesfully retrieve a value:
     key with address90FDB51C13CE085CE7F9C0DA8683B7327711B064not found: key not found
  

  2) Contract: Test Storage Contract
       "after each" hook: after test for "should succesfully retrieve a value":
     failed to fetch header by number (latest): rpc error: code = NotFound desc = rpc error: code = NotFound desc = block bloom not found: key not found
  



error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

```

__Additional info:__ 

1. The 3 test cases are:
 * Deploy a smart contract, Expect it to be deployed
 * Write a value (888) and store it on the deployed smart contract. Expect the tx to be successful
 * Read the value from smart contract by making a call (not transaction). Expect the value to be 888

**Test Source Code:**
```
contract('Test Storage Contract', async function (accounts) {

  let storageInstance;

  before(function () {
    console.log(`Using Accounts (${accounts.length}): \n${accounts.join('\n')}`);
    console.log('==========================\n');
  })

  it('should deploy Stroage contract', async function () {
    storageInstance = await Storage.new();
    console.log(`Deployed Storage at: ${storageInstance.address}`);
    expect(storageInstance.address).not.to.be.undefined;
  });

  it('should succesfully stored a value', async function () {
    const tx = await storageInstance.store(888);
    console.log(`Stored value 888 by tx: ${tx.tx}`);
    expect(tx.tx).not.to.be.undefined;
  });

  it('should succesfully retrieve a value', async function () {
    const value = await storageInstance.retrieve();
    expect(value.toString()).to.equal('888');
  });

})
```

**Comment by yijiasu-crypto** [comment]
The problematic part of this issue is:

* In second round of test, if Test 1 and 2 are passed, the Test 3 should be passed as well. Since it doesn't change anything on the blockchain but only reading the value, it should be successful all the time
* If removing the parameter `--keyring-backend test` result in the `ethermintd` node unable to read the keyring and load it on chain initialization, then all tests should be FAILED as well. Test 1 & 2 passed means:
  * There's balance in genesis accounts
  * The blockchain node accepts the transaction and record it on block ( this can be proved by calling `web3.eth.getCode("0xedF82B7bF42006dDD309cB2DEC789316407b3c55")`, we can see the returned bytecode)
 
In such scenario, there's no reason to believe that the Test 3# will fail.

**Comment by yijiasu-crypto** [comment]
@leejw51crypto Can you assign yourself on this since it seems I can't change the assignee

**Comment by leejw51crypto** [comment]
started, thanks for nice catch @yijiasu-crypto

**Comment by yihuang** [comment]
Is there a simpler description of the issue? I can access the repo `crypto.com-evm-chain-test` currently.

**Comment by yihuang** [comment]
https://github.com/tharsis/ethermint/issues/111
Possibly same issue as this, always use `--keyring-backend test` in current status.

**Comment by leejw51crypto** [comment]
with --keyring-backend file or os
, i entered keys,
sending tx working fine

i'll try solidity now

**Comment by leejw51crypto** [comment]
after this merged
https://github.com/tharsis/ethermint/pull/236

let's check again~

**Comment by yijiasu-crypto** [comment]
This can be considered as resolved since I ran two unit tests against #236 all passed.
